File: frontend/MediaFlow/MediaFlow/Models/WorkerServer.swift
import Foundation

struct PathMapping: Codable, Identifiable {
    var id = UUID()
    var sourcePrefix: String
    var targetPrefix: String

    enum CodingKeys: String, CodingKey {
        case sourcePrefix, targetPrefix
    }
}

struct WorkerServer: Identifiable, Codable {
    let id: Int
    let name: String
    let hostname: String
    var port: Int = 22
    var sshUsername: String?
    var role: String = "transcode"
    var gpuModel: String?
    var cpuModel: String?
    var cpuCores: Int?
    var ramGb: Double?
    var hwAccelTypes: [String]?
    var maxConcurrentJobs: Int = 1
    var status: String = "offline"
    var lastHeartbeatAt: String?
    var hourlyCost: Double?
    var isLocal: Bool = false
    var isEnabled: Bool = true
    var workingDirectory: String = "/tmp/mediaflow"
    var pathMappings: [PathMapping]?
    var activeJobs: Int = 0
    var performanceScore: Double?
    var lastBenchmarkAt: String?
    // Cloud GPU fields
    var cloudProvider: String?
    var cloudInstanceId: String?
    var cloudPlan: String?
    var cloudRegion: String?
    var cloudCreatedAt: String?
    var cloudAutoTeardown: Bool = true
    var cloudIdleMinutes: Int = 30
    var cloudStatus: String?    // "creating", "bootstrapping", "active", "destroying", "destroyed"

    var isCloud: Bool { cloudProvider != nil }
}

struct ServerStatus: Codable {
    let id: Int
    let name: String
    let status: String
    var cpuPercent: Double?
    var gpuPercent: Double?
    var ramUsedGb: Double?
    var ramTotalGb: Double?
    var gpuTemp: Double?
    var fanSpeed: Int?
    var activeJobs: Int = 0
    var queuedJobs: Int = 0
    var uptimePercent: Double?
    var uploadMbps: Double?
    var downloadMbps: Double?
    var performanceScore: Double?
}

struct BenchmarkResult: Identifiable, Codable {
    let id: Int
    let workerServerId: Int
    var uploadMbps: Double?
    var downloadMbps: Double?
    var latencyMs: Double?
    var testFileSizeBytes: Int = 200_000_000
    var status: String
    var errorMessage: String?
    var startedAt: String?
    var completedAt: String?
    var createdAt: String?
}

struct BenchmarkTriggerResponse: Codable {
    let status: String
    var benchmarkId: Int?
    let message: String
}

struct ServerPickerItem: Identifiable, Codable {
    let id: Int
    let name: String
    let status: String
    var performanceScore: Double?
    var activeJobs: Int = 0
    var queuedJobs: Int = 0
    var maxConcurrentJobs: Int = 1
    var cpuModel: String?
    var gpuModel: String?
    var uploadMbps: Double?
    var downloadMbps: Double?
    var isLocal: Bool = false
}

struct ServerEstimateResponse: Codable {
    let serverId: Int
    let serverName: String
    var estimatedSeconds: Int?
    var estimatedDisplay: String = "--"
    var basedOnJobs: Int = 0
}

struct ProvisionTriggerResponse: Codable {
    let status: String
    let message: String
}

struct ProvisionProgress {
    let serverId: Int
    let progress: Int
    let step: String
    let message: String
}

struct CloudDeployProgress {
    let serverId: Int
    let step: String
    let progress: Int
    let message: String
}

// MARK: - Cloud GPU Models

struct CloudPlanInfo: Identifiable, Codable {
    var id: String { planId }
    let planId: String
    let gpuModel: String
    let vcpus: Int
    let ramMb: Int
    let gpuVramGb: Int
    let monthlyCost: Double
    let hourlyCost: Double
    let regions: [String]
}

struct CloudDeployRequest: Codable {
    let plan: String
    let region: String
    let idleMinutes: Int
    let autoTeardown: Bool
}

struct CloudDeployResponse: Codable {
    let status: String
    let serverId: Int
    let message: String
}

struct CloudCostRecordResponse: Codable, Identifiable {
    let id: Int
    var workerServerId: Int?
    var jobId: Int?
    let cloudProvider: String
    let cloudInstanceId: String
    let cloudPlan: String
    let hourlyRate: Double
    let startTime: String
    var endTime: String?
    var durationSeconds: Double?
    var costUsd: Double?
    let recordType: String
}

struct CloudCostSummary: Codable {
    let currentMonthTotal: Double
    let activeInstanceRunningCost: Double
    let monthlyCap: Double
    let instanceCap: Double
    let records: [CloudCostRecordResponse]
}

struct CloudSettingsResponse: Codable {
    let apiKeyConfigured: Bool
    let defaultPlan: String
    let defaultRegion: String
    let monthlySpendCap: Double
    let instanceSpendCap: Double
    let defaultIdleMinutes: Int
}

struct CloudSettingsUpdate: Codable {
    var vultrApiKey: String?
    var defaultPlan: String?
    var defaultRegion: String?
    var monthlySpendCap: Double?
    var instanceSpendCap: Double?
    var defaultIdleMinutes: Int?
}

File: backend/app/models/worker_server.py
from sqlalchemy import Column, Integer, String, Float, Boolean, JSON, DateTime, func
from sqlalchemy.orm import relationship
from app.database import Base


class WorkerServer(Base):
    __tablename__ = "worker_servers"

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(100), nullable=False)
    hostname = Column(String(255), nullable=False)
    port = Column(Integer, default=22)
    ssh_username = Column(String(100), nullable=True)
    ssh_key_path = Column(String(500), nullable=True)
    role = Column(String(50), default="transcode")
    gpu_model = Column(String(100), nullable=True)
    cpu_model = Column(String(100), nullable=True)
    cpu_cores = Column(Integer, nullable=True)
    ram_gb = Column(Float, nullable=True)
    hw_accel_types = Column(JSON, nullable=True)
    max_concurrent_jobs = Column(Integer, default=1)
    status = Column(String(20), default="offline", index=True)
    last_heartbeat_at = Column(DateTime, nullable=True)
    hourly_cost = Column(Float, nullable=True)
    is_local = Column(Boolean, default=False)
    is_enabled = Column(Boolean, default=True)
    working_directory = Column(String(500), default="/tmp/mediaflow")
    path_mappings = Column(JSON, nullable=True, default=[])
    # Format: [{"source_prefix": "/share/ZFS18_DATA/", "target_prefix": "/Volumes/MediaNAS/"}]
    performance_score = Column(Float, nullable=True)  # 0-100, from benchmarks
    last_benchmark_at = Column(DateTime, nullable=True)
    consecutive_failures = Column(Integer, default=0)
    provision_log = Column(String, nullable=True)  # JSON result of last provisioning
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    transcode_jobs = relationship("TranscodeJob", back_populates="worker_server")
    job_logs = relationship("JobLog", back_populates="worker_server")
    benchmarks = relationship("ServerBenchmark", back_populates="worker_server", cascade="all, delete-orphan")
